File: utilities/screenshot_website.rb
#!/usr/bin/env ruby

require 'fileutils'
require 'front_matter_parser'

chrome_path = '/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome'
website_image_dir = 'assets/images/websites'
resize_image_dir = 'assets/images/resized'
website_image_extension='jpg'
window_width = 1024
window_height = 640
# The delay (in milliseconds) to wait before taking the screenshot
screenshot_delay = 1000

# Create website image directory if it doesn't already exist
FileUtils.mkdir_p(website_image_dir)
# Remove existing directory containing resized images
FileUtils.remove_dir(resize_image_dir, force=true)

# Generate screenshot for the specified file path(s); otherwise, default to
# every file in the _websites directory
website_config_files = ARGV
if website_config_files.empty?
  website_config_files = Dir.glob('_websites/*.md')
end

# Generate screenshot for each portfolio website that has configuration
website_config_files.each do |website_config_file|

  website_name = File.basename(website_config_file, '.md')
  puts "generating screenshot for #{website_name}"

  parsed = FrontMatterParser::Parser.parse_file(website_config_file)
  system("""#{chrome_path} \
    --headless \
    --disable-gpu \
    --disk-cache-dir=/dev/null \
    --window-size=#{window_width},#{window_height} \
    --hide-scrollbars \
    --virtual-time-budget=#{screenshot_delay} \
    --screenshot=#{website_image_dir}/#{website_name}.#{website_image_extension} \
    #{parsed['direct_url']}
  """)

end

File: utilities/screenshot-website.ts
// To run this utility, run `esr utilities/screenshot-website.ts` from the root
// project directory on the CLI
import fs from 'fs';
import glob from 'glob-promise';
import path from 'path';
import puppeteer from 'puppeteer';
import type { WebsiteEntry } from '../src/routes/types.d';

const websiteImageDir = 'static/images/websites';
const websiteImageExtension = 'jpg';
const websiteImageQuality = 85;
const windowWidth = 1024;
const windowHeight = 640;

// Create website image directory if it doesn't already exist
function createWebsiteImageDirectory(): Promise<string | undefined> {
  return fs.promises.mkdir(websiteImageDir, { recursive: true });
}

async function generateScreenshots(websiteConfigFilePaths: string[]): Promise<void> {
  await createWebsiteImageDirectory();
  console.log('launching browser...');
  const browser = await puppeteer.launch();

  // Generate screenshot for each portfolio website that has configuration
  await Promise.all(
    websiteConfigFilePaths.map(async (websiteConfigFilePath) => {
      const websiteName = path.basename(websiteConfigFilePath, '.json');

      const websiteConfigFileContents = await fs.promises.readFile(websiteConfigFilePath, 'utf8');

      const websiteEntry = JSON.parse(websiteConfigFileContents) as WebsiteEntry;
      const websiteImagePath = path.join(
        websiteImageDir,
        `${websiteName}.${websiteImageExtension}`
      );

      const page = await browser.newPage();
      await page.setViewport({
        width: windowWidth,
        height: windowHeight
      });
      await page.goto(websiteEntry.direct_url, {
        waitUntil: ['load', 'networkidle0']
      });

      return page
        .screenshot({
          path: websiteImagePath,
          type: 'jpeg',
          quality: websiteImageQuality,
          captureBeyondViewport: false
        })
        .then(() => {
          console.log(`generated screenshot for ${websiteName}`);
        })
        .catch((error) => {
          console.error(error);
          return error;
        });
    })
  );

  console.log('closing browser...');
  await browser.close();
}

async function main(): Promise<void> {
  // Generate screenshot for the specified file path(s); otherwise, default to
  // every file in the /src/websites directory
  let websiteConfigFilePaths = process.argv.slice(2);
  if (websiteConfigFilePaths.length === 0) {
    websiteConfigFilePaths = await glob.promise('src/websites/*.json');
  }
  generateScreenshots(websiteConfigFilePaths);
}
main();
